### PR TITLE
Log mismatch between NUM_CPU and actual cpu usage

### LINKED ISSUE #8721: Detect discrepancies between NUM_CPU and what has actually happened, and log it if something is found

### PR DESCRIPTION
**Pull request by berland**
**Issue**
Resolves #8721 

**Approach**
calculate and log.

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'pytest tests/unit_tests -n logical -m "not integration_test"'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/ensemble_evaluator/evaluator.py [hunk 1]
```diff
@@ -161,12 +162,34 @@ async def _stopped_handler(self, events: Sequence[EnsembleSucceeded]) -> None:
...
 
         max_memory_usage = -1
-        for fm_step in self.ensemble.snapshot.get_all_fm_steps().values():
+        for (real_id, _), fm_step in self.ensemble.snapshot.get_all_fm_steps().items():
+            # Infer max memory usage
             memory_usage = fm_step.get(ids.MAX_MEMORY_USAGE) or "-1"
             max_memory_usage = max(int(memory_usage), max_memory_usage)
+
+            # Detect over-spending of CPU
+            num_cpu = self.ensemble.reals[int(real_id)].num_cpu
```
Comment (eivindjahren) [inline src/ert/ensemble_evaluator/evaluator.py:171]: We should probably extract this into a function.

**Review by eivindjahren — CHANGES_REQUESTED** [review]
just think a quick extract method will help us in the future here.

File: tests/ert/unit_tests/ensemble_evaluator/test_ensemble_evaluator.py [hunk 2]
```diff
@@ -465,3 +474,26 @@ async def test_ensure_multi_level_events_in_order(evaluator_to_use):
...
+    start: datetime.datetime,
+    end: datetime.datetime,
+    cpu_seconds: float,
+):
+    message = detect_overspent_cpu(
+        num_cpu, "dummy", FMStepSnapshot(start=start, end=end, cpu_seconds=cpu_seconds)
+    )
+    duration = (end - start).total_seconds()
+    print(f"{duration=} {cpu_seconds/duration=} {num_cpu=}")
+    if duration > 0 and cpu_seconds / duration > num_cpu:
```
Comment (andreas-el) [inline tests/ert/unit_tests/ensemble_evaluator/test_ensemble_evaluator.py:496]: So if duration is `1`, and cpu_seconds is `1`, and incidentally num_cpu is `1` .. ?

**Review by andreas-el — CHANGES_REQUESTED** [review]
This looks really promising, but you'd need to look into the failing tests.

File: src/ert/ensemble_evaluator/evaluator.py [hunk 3]
```diff
@@ -161,12 +163,20 @@ async def _stopped_handler(self, events: Sequence[EnsembleSucceeded]) -> None:
...
-        for fm_step in self.ensemble.snapshot.get_all_fm_steps().values():
+        for (real_id, _), fm_step in self.ensemble.snapshot.get_all_fm_steps().items():
+            # Infer max memory usage
             memory_usage = fm_step.get(ids.MAX_MEMORY_USAGE) or "-1"
             max_memory_usage = max(int(memory_usage), max_memory_usage)
+
+            if cpu_message := detect_overspent_cpu(
+                self.ensemble.reals[int(real_id)].num_cpu, real_id, fm_step
+            ):
+                logger.warning(cpu_message)
```
Comment (eivindjahren) [inline src/ert/ensemble_evaluator/evaluator.py:174]: Seems like the tests are not hitting the warning line currently.

**Comment by berland** [comment]
Fixed the test now.

File: tests/ert/unit_tests/ensemble_evaluator/test_ensemble_evaluator.py (hunk 2 above)
Comment (berland) [reply tests/ert/unit_tests/ensemble_evaluator/test_ensemble_evaluator.py:500]: then no warning should be printed.

**Comment by berland** [comment]
We could perhaps allow some slack, say 5%, but that is more important when/if we flash this to the users and not just logging it.

File: tests/ert/unit_tests/ensemble_evaluator/test_ensemble_evaluator.py (hunk 2 above)
Comment (andreas-el) [reply tests/ert/unit_tests/ensemble_evaluator/test_ensemble_evaluator.py:500]: Indeed. 🤦

**Review by andreas-el — APPROVED** [review]
💯

**Comment by berland** [comment]
@eivindjahren you also need to approve since you have requested changes :)

**Pull request merged**